tensorflow/python/ops/gradients.py

```diff
@@ -829,3 +829,45 @@ def _hessian_vector_product(ys, xs, v):
 
   # Second backprop
   return gradients(elemwise_products, xs)
+
+
+def hessian(ys, x, name="hessian", colocate_gradients_with_ops=False, 
+            gate_gradients=False, aggregation_method=None):
+  """Constructs the Hessian of sum of `ys` with respect to `x`.
+
+  `hessian()` adds ops to the graph to output the Hessian of `ys` with 
+  respect to `x` and returns a `Tensor`.
+
+  Args:
+    ys: A `Tensor` or list of tensors to be differentiated.
+    x: A `Tensor` to be used for differentiation.
+    name: Optional name to use for grouping all the gradient ops together.
+      defaults to 'gradients'.
+    colocate_gradients_with_ops: If True, try colocating gradients with
+      the corresponding op.
+    gate_gradients: If True, add a tuple around the gradients returned
+      for an operation.  This avoids some race conditions.
+    aggregation_method: Specifies the method used to combine gradient terms.
+      Accepted values are constants defined in the class `AggregationMethod`.
+
+  Returns:
+    The hessian of `sum(ys)` with respect to `x`.
+
+  Raises:
+    LookupError: if one of the operations between `x` and `ys` does not
+      have a registered gradient function.
+    ValueError: if the arguments are invalid
+  """
+  ndims = x.get_shape().ndims
+  if ndims is not None and ndims != 1:
+    raise ValueError("Hessians can only be computed with respect to one-dimensional tensors")
+  kwargs = {
+    'colocate_gradients_with_ops': colocate_gradients_with_ops,
+    'gate_gradients': gate_gradients,
+    'aggregation_method': aggregation_method
+  }
+  # Compute the regular gradients
+  _gradients, = gradients(ys, x, **kwargs)
+  # Apply the gradients again
+  _hess = [gradients(_gradient, x, **kwargs)[0] for _gradient in array_ops.unpack(_gradients)]
```

Comment:
Okay, I now appreciate a bit more about the subtlety of packing/unpacking now that I'm reviewing the code :).  

As far as I understand, _gradients is a single element Tensor containing the sum(dy/dx), and you're unpacking the tensor along the first dimension (which is guaranteed to be a 1D tensor) into its scalar components, and then performing gradients again for each element in the vector with respect to 'x'.  Finally, you pack the results together.

If this is correct, can you add some comments to describe what is being done here?  Future readers will greatly appreciate it!